### PR TITLE
Child removal

### PR DESCRIPTION
**Pull request by kanes115**
This PR tests and corrects pr #169.

TODO:
- [x] Test situation in which playbackbuffer has buffers from different elements.
- [x] get rid of sleeps after rebasing with new test API branch. (There was no point to do this on old test api).
- [x] Remove list args where only one arg is passed (Message)

### CODE REVIEW

**Comment by kanes115** [comment]
All suggestions regarding tests (if they're sensible, well structured etc) welcome.

File: test/support/child_removal_test/filter.ex [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+  def_options demand_generator: [
+                type: :function,
+                spec: (pos_integer -> non_neg_integer),
+                default: &__MODULE__.default_demand_generator/1
+              ],
+              target: [type: :pid],
+              playing_delay: [type: :integer, default: 0],
+              ref: [type: :any, default: nil],
+              two_input_pads: [type: :boolean, default: false],
```
Comment (mat-hek) [inline test/support/child_removal_test/filter.ex]: once you switch to dynamic input you can get rid of this option in favour of `handle_pad_added`

File: lib/membrane/core/element/playback_buffer.ex [hunk 2]
```diff
@@ -77,6 +77,16 @@ defmodule Membrane.Core.Element.PlaybackBuffer do
 
   def eval(state), do: {:ok, state}
 
+  def flush_for_pad(%__MODULE__{q: q} = buf, pad_ref) do
+    q
+    |> Enum.filter(fn
+      {_, [_, ^pad_ref]} -> false
```
[the remark this answers is not in the document]
Comment (kanes115) [reply lib/membrane/core/element/playback_buffer.ex]: Why did we use a list here in the first place?

File: test/membrane/integration/child_removal_test.exs [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    assert Pipeline.stop(pid) == :ok
+    assert_receive :pipeline_stopped, 500
+  end
+
+  defp assert_pid_dead(pid) do
+    assert_receive {:DOWN, _, :process, ^pid, :normal}
+  end
+
+  defp assert_pid_alive(pid) do
+    refute_receive {:DOWN, _, :process, ^pid, _}
```
[the remark this answers is not in the document]
Comment (kanes115) [reply test/membrane/integration/child_removal_test.exs]: You mean why do we assert that process is alive?

File: test/membrane/integration/child_removal_test.exs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    stop_pipeline(pipeline_pid)
+  end
+
+  #############
+  ## HELPERS ##
+  #############
+
+  defp stop_pipeline(pid) do
+    assert Pipeline.stop(pid) == :ok
+    assert_receive :pipeline_stopped, 500
```
[the remark this answers is not in the document]
Comment (kanes115) [reply test/membrane/integration/child_removal_test.exs]: If I understand correctly, I cannot as I don't use the testing Pipeline.

File: test/membrane/integration/child_removal_test.exs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    ChildRemovalTest.Pipeline.remove_child(pipeline_pid, :filter1)
+
+    assert_receive {:element_shutting_down, ^filter_pid1}
+    assert_pid_dead(filter_pid1)
+    assert Process.alive?(filter_pid2)
+
+    stop_pipeline(pipeline_pid)
+  end
+
+  @doc """
```
[the remark this answers is not in the document]
Comment (kanes115) [reply test/membrane/integration/child_removal_test.exs]: I find those test cases quite complex. The way this test fills playback buffer up, IMO, needs explanation.

I believe that documentation under Integration tests is sometimes (or one should even say oftentimes) very useful.

File: test/membrane/integration/child_removal_test.exs [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    ChildRemovalTest.Pipeline.remove_child(pipeline_pid, :filter1)
+
+    assert_receive {:element_shutting_down, ^filter_pid1}
+    assert_pid_dead(filter_pid1)
+    assert_pid_alive(filter_pid2)
+
+    stop_pipeline(pipeline_pid)
+  end
+
+  @doc """
```
[the remark this answers is not in the document]
Comment (kanes115) [reply test/membrane/integration/child_removal_test.exs]: [same answer](https://github.com/membraneframework/membrane-core/pull/173#discussion_r306769022)

File: test/membrane/integration/child_removal_test.exs [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+  defp get_filter_pid(ref) do
+    assert_receive {:filter_pid, ^ref, pid}
+    Process.monitor(pid)
+    pid
+  end
+
+  defp prepared_to_playing_delay, do: 300
+
+  # Checks if there is at least one buffer element for each pad specified in `for_pads`
```
[the remark this answers is not in the document]
Comment (kanes115) [reply test/membrane/integration/child_removal_test.exs]: ?

File: test/support/child_removal_test/filter.ex [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  def_input_pad :input1, demand_unit: :buffers, caps: :any
+
+  def_input_pad :input2, demand_unit: :buffers, caps: :any
+
+  def_options demand_generator: [
+                type: :function,
+                spec: (pos_integer -> non_neg_integer),
+                default: &__MODULE__.default_demand_generator/1
+              ],
+              target: [type: :pid],
```
[the remark this answers is not in the document]
Comment (kanes115) [reply test/support/child_removal_test/filter.ex]: I cannot notify the watcher from `handle_init`, can I?

File: lib/membrane/core/element/action_handler.ex [hunk 9]
```diff
@@ -209,7 +209,7 @@ defmodule Membrane.Core.Element.ActionHandler do
         pad_data
 
       state = handle_buffer(pad_ref, mode, other_demand_unit, buffers, state)
-      Message.send(pid, :buffer, [buffers, other_ref])
+      Message.send(pid, :buffer, buffers, from_pad: other_ref)
```
Comment (kanes115) [inline lib/membrane/core/element/action_handler.ex]: `from_pad` might not be the best name...

Comment (mat-hek) [reply lib/membrane/core/element/action_handler.ex]: Agreed, I'd rather call this `for_pad`

File: test/support/child_removal_test/filter.ex (hunk 8 above)
[the remark this answers is not in the document]
Comment (mat-hek) [reply test/support/child_removal_test/filter.ex]: No, but you can do that upon `handle_prepare` instead. Or if it's needed we can easily support that

File: test/support/child_removal_test/pipeline.ex [hunk 10]
```diff
@@ -11,6 +11,9 @@ defmodule Membrane.Support.ChildRemovalTest.Pipeline do
                                      /
                     extra_source ___/
 
+  This pipeline also makes children aware of their names. They can reach their
```
Comment (mat-hek) [inline test/support/child_removal_test/pipeline.ex]: why is this needed?

File: test/support/child_removal_test/filter.ex [hunk 11]
```diff
@@ -35,23 +34,24 @@ defmodule Membrane.Support.ChildRemovalTest.Filter do
               target: [type: :pid],
               playing_delay: [type: :integer, default: 0],
               ref: [type: :any, default: nil],
-              two_input_pads: [type: :boolean, default: false],
-              sof_sent?: [type: :boolean, default: false],
-              input1_deactivated: [type: :boolean, default: false]
+              sof_sent?: [type: :boolean, default: false]
```
Comment (mat-hek) [inline test/support/child_removal_test/filter.ex]: why is this needed?

Comment (kanes115) [reply test/support/child_removal_test/filter.ex]: To know that we already sent start of stream event and be able to only forward one start of stream. I made a mistake in the name of this option. Should be `sos_sent?`.

File: test/membrane/integration/child_removal_test.exs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mat-hek) [reply test/membrane/integration/child_removal_test.exs]: So you can use it :)

File: test/membrane/integration/child_removal_test.exs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mat-hek) [reply test/membrane/integration/child_removal_test.exs]: I mean why `assert Process.alive?(pid)` is not enough

File: test/membrane/integration/child_removal_test.exs (hunk 7 above)
[the remark this answers is not in the document]
Comment (mat-hek) [reply test/membrane/integration/child_removal_test.exs]: "Checks if there is at least one item in the playback buffer for each pad specified in `for_pads`" would be more descriptive IMHO

File: test/support/child_removal_test/filter.ex (hunk 11 above)
Comment (mat-hek) [reply test/support/child_removal_test/filter.ex]: but why this has to be passed as an option, and not just stored in the state? BTW it even doesn't need to be stored in the state, as it's available in the context under `context.pads.pad_name.start_of_stream?`

File: test/support/child_removal_test/filter.ex [hunk 12]
```diff
@@ -56,8 +52,8 @@ defmodule Membrane.Support.ChildRemovalTest.Filter do
 
   @impl true
   def handle_prepared_to_playing(_ctx, %{playing_delay: 0} = state) do
-    send(state.target, {:playing, self()})
-    {:ok, state}
+    # send(state.target, {:playing, self()})
```
Comment (mat-hek) [inline test/support/child_removal_test/filter.ex]: dupa

File: test/support/child_removal_test/pipeline.ex [hunk 13]
```diff
@@ -63,8 +57,7 @@ defmodule Membrane.Support.ChildRemovalTest.Pipeline do
   end
 
   @impl true
-  def handle_prepared_to_playing(%{target: target} = state) do
-    send(target, {:playing, self()})
+  def handle_prepared_to_playing(state) do
```
Comment (mat-hek) [inline test/support/child_removal_test/pipeline.ex]: can be removed

File: test/membrane/integration/child_removal_test.exs (hunk 4 above)
[the remark this answers is not in the document]
Comment (kanes115) [reply test/membrane/integration/child_removal_test.exs]: I cannot as testing pipeline only allows for flat structure.
